README.md
# ods-toast web component

The `<ods-toast>` component is custom functional web component for the purpose of exposing a user feedback interface.

## Docs

All information regarding Project Setup, Technical Details, Tests and information regarding auro web components can be found in the [doc site](https://github.com/AlaskaAirlines/OrionStatelessComponents__docs/tree/master/docs).

## Install

[![Build Status](https://img.shields.io/travis/AlaskaAirlines/ods-toast?branch=master&style=for-the-badge)](https://travis-ci.org/github/AlaskaAirlines/ods-toast)
[![See it on NPM!](https://img.shields.io/npm/v/@alaskaairux/ods-toast?style=for-the-badge&color=orange)](https://www.npmjs.com/package/@alaskaairux/ods-toast)
[![License](https://img.shields.io/npm/l/@alaskaairux/ods-toast?color=blue&style=for-the-badge)](https://www.apache.org/licenses/LICENSE-2.0)

```shell
$ npm i @alaskaairux/ods-toast
```

### Define dependency in project component

Define the component dependency within each component that is using the `<ods-toast>` component.

```javascript
import "@alaskaairux/ods-toast";
```

### Toaster.js

Use the bundled [Toaster.js](https://auro.aslakaair.com/components/orion/toast/toaster) tool to display and manage toasts in a given interface. This dependency isn't built into the scope of the component in order to allow for consuming projects to bundle it with other project dependencies as needed.

Toaster.js is an optional--but highly recommended--utility class containing logic to manage toasts. If there is a compelling reason not to use Toaster.js, please submit an issue with the intent to resolve it.

### Swipe.js

Swipe.js is an independent dependency that will only be loaded when the user's device is detected to support [touchstart](https://developer.mozilla.org/en-US/docs/Web/API/Element/touchstart_event).

### Toaster.css / .scss

The animation of `<ods-toast>` is to be installed as an external resource.

| File | Type | Import path |
|---|---|---|
| toaster.css | CSS | @alaskaairux/ods-toast/dist/toaster.css |
| toaster.scss | Sass | @alaskaairux/ods-toast/dist/toaster.scss |

### Orion Icons dependency

The `<ods-toast>` custom element has a dependency on the Auro Icons library. Be sure to read the [Icons tokens install instructions](https://auro.aslakaair.com/icons/install) for additional styling support.

### Design Token CSS Custom Property dependency

The use of any ODS Component has a dependency on the [ODS Design Tokens](https://auro.alaskaair.com/getting-started/developers/design-tokens).

### CSS Custom Property fallbacks

CSS Custom Properties are not supported in older browsers. For this, fallback properties are pre-generated and included with the npm. Any update to the Orion Design Tokens will be immediately reflected with browsers that support CSS Custom Properties, legacy browsers will require updated components with pre-generated fallback properties.

**Reference component in HTML**

```html
<ods-toast title="This is the title"></ods-toast>
```

## Element ods-toast

### Toast use cases

The `<ods-toast>` element should be used in situations where:

- alert of successful actions is passive
- message does not require interaction from user
- messages, if missed by auto dismiss, is not crucial to the users interaction

### Properties:

| Attribute | Value type   | Description                                             |
| --------- | ------------ | ------------------------------------------------------- |
| title     | string       | large bold-faced text at the top of the toast component |
| message   | string, null | smaller text that appears below the title               |
| clickCallback | function | to allow for external events to be bound within the scope of the shadow DOM |


### API Code Examples

#### Default Classic Toast

```html
<ods-toast title="This is a title"></ods-toast>
```

```html
<ods-toast title="This is a title" message="This is a message"></ods-toast>
```

#### Classic Toast with slotting

```html
<ods-toast title="This is a title" message="This is a message">
  <ods-hyperlink class="util_paddingLeft--none" href="#">Learn more</ods-hyperlink>
</ods-toast>
```

##### Loading Dynamic Toasts

Loading a dynamic experience with `<ods-toast>` requires the use of Toaster.js and Swipe.js and follows a slightly different API then that of the static web component.

The following example illustrates how you could implement a use case:

```js
import "@alaskaairux/ods-toast/dist/ods-toast";
import Swipe from "@alaskaairux/ods-toast/dist/swipe.js";
import Toaster from "@alaskaairux/ods-toast/dist/toaster";
import "@alaskaairux/ods-toast/dist/toaster.css";

window.Swipe = Swipe;
const toastDuration = 7000;
const initialToastDelay = 2000;
const toasts = new Toaster(toastDuration);
setTimeout(() => {
  toasts.add("Your status has been updated");
  toasts.add("Flight information updated", "Thank you, Your seat preference has been saved");
  toasts.add("Wahoo!", "Your preferences have been updated", "Oops? <ods-hyperlink class='util_paddingLeft--none' href='/'>UNDO</ods-hyperlink>");
}, initialToastDelay);
```

Each instance of `toasts.add()` will add the `<ods-toast>` custom element to the view.

> ods-hyperlink example styling; requires @alaskaairux/ods-hyperlink: ^1.0.6

> The classes `util_paddingLeft--none` and `util_paddingRight--none` can be placed on the `ods-hyperlink` element using the class attribute.


## Development

In order to develop against this project, if you are not part of the core team, you will be required to fork the project prior to submitting a pull request.

Please be sure to review the [contribution guidelines](https://auro.alaskaair.com/getting-started/developers/contributing) for this project. Please make sure to **pay special attention** to the **conventional commits** section of the document.

### Start development environment

Once the project has been cloned to your local resource and you have installed all the dependencies you will need to open three different shell sessions. One is for the Gulp tasks, the second is for a series of npm tasks and the last is to run the Polymer server.

```shell
// shell terminal one
$ gulp dev

// shell terminal two
$ npm run dev

// shell terminal three
polymer serve
```


CHANGELOG.md
# Semantic Release Automated Changelog

## [1.1.3](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/compare/v1.1.2...v1.1.3) (2020-07-17)


### Bug Fixes

* update docs and package versions ([d2f6ecf](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/commit/d2f6ecfb7e4294729e82fecb4fb5d36f1040e75a))

## [1.1.2](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/compare/v1.1.1...v1.1.2) (2020-03-30)


### Bug Fixes

* [#7](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/issues/7) address fallback CSS ([2ff2926](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/commit/2ff2926))

## [1.1.1](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/compare/v1.1.0...v1.1.1) (2019-11-19)


### Bug Fixes

* updated icon folder location ([44c4f82](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/commit/44c4f82))

# [1.1.0](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/compare/v1.0.7...v1.1.0) (2019-11-18)


### Features

* updated Orion icon dep to latest name ([9eb530d](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/commit/9eb530d))

## [1.0.7](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/compare/v1.0.6...v1.0.7) (2019-10-18)


### Bug Fixes

* add barrel file to simplify import statements ([fd0bd4d](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/commit/fd0bd4d))

## [1.0.6](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/compare/v1.0.5...v1.0.6) (2019-09-27)


### Bug Fixes

* docs and package.json update ([11844b1](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/commit/11844b1))

## [1.0.5](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/compare/v1.0.4...v1.0.5) (2019-09-06)


### Bug Fixes

* update entrance animation timing ([35f06dc](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/commit/35f06dc))

## [1.0.4](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/compare/v1.0.3...v1.0.4) (2019-09-06)


### Bug Fixes

* add back X on mobile; increase tap/swipe target ([0575e9b](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/commit/0575e9b))

## [1.0.3](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/compare/v1.0.2...v1.0.3) (2019-09-06)


### Bug Fixes

* adjust mobile interface and increase tap area ([231a0d6](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/commit/231a0d6))

## [1.0.2](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/compare/v1.0.1...v1.0.2) (2019-09-06)


### Bug Fixes

* include CSS/SCSS resources, hide X on mobile ([511f8b0](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/commit/511f8b0))

## [1.0.1](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/compare/v1.0.0...v1.0.1) (2019-09-06)


### Bug Fixes

* [#5](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/issues/5) address iOS platform issue; update docs and add images to README ([c8da7bb](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/commit/c8da7bb))

# 1.0.0 (2019-09-04)


### Bug Fixes

* address post review issues ([806398e](https://github.com/AlaskaAirlines/OrionStatelessComponents__ods-toast/commit/806398e))


.eslintrc.js
module.exports = {
  "env": {
      "browser": true,
      "es6": true
  },
  "extends": [
      "eslint:recommended",
      "plugin:lit/recommended"
  ],
  "globals": {
      "Atomics": "readonly",
      "SharedArrayBuffer": "readonly"
  },
  "parserOptions": {
      "ecmaVersion": 2018,
      "sourceType": "module"
  },
  "plugins": [
      "lit"
  ],
  "rules": {
      "accessor-pairs": "error",
      "array-bracket-newline": "error",
      "array-bracket-spacing": "error",
      "array-callback-return": "error",
      "array-element-newline": "error",
      "arrow-body-style": "error",
      "arrow-parens": "error",
      "arrow-spacing": [
          "error",
          {
              "after": true,
              "before": true
          }
      ],
      "block-scoped-var": "error",
      "block-spacing": "error",
      "brace-style": "error",
      "callback-return": "error",
      "camelcase": "error",
      "capitalized-comments": "off",
      "class-methods-use-this": "off",
      "comma-dangle": "off",
      "comma-spacing": [
          "error",
          {
              "after": true,
              "before": false
          }
      ],
      "comma-style": [
          "error",
          "last"
      ],
      "complexity": "error",
      "computed-property-spacing": "error",
      "consistent-return": "error",
      "consistent-this": "error",
      "curly": "error",
      "default-case": "error",
      "default-param-last": "error",
      "dot-location": "error",
      "dot-notation": "error",
      "eol-last": "off",
      "eqeqeq": "error",
      "func-call-spacing": "error",
      "func-name-matching": "error",
      "func-names": "off",
      "func-style": "error",
      "function-paren-newline": "error",
      "generator-star-spacing": "error",
      "global-require": "error",
      "grouped-accessor-pairs": "error",
      "guard-for-in": "error",
      "handle-callback-err": "error",
      "id-blacklist": "error",
      "id-length": "error",
      "id-match": "error",
      "implicit-arrow-linebreak": "error",
      "indent": "off",
      "indent-legacy": "off",
      "init-declarations": "error",
      "jsx-quotes": "error",
      "key-spacing": "off",
      "keyword-spacing": [
          "error",
          {
              "after": true,
              "before": true
          }
      ],
      "line-comment-position": "error",
      "linebreak-style": [
          "error",
          "unix"
      ],
      "lines-around-comment": "error",
      "lines-around-directive": "error",
      "lines-between-class-members": [
          "error",
          "always"
      ],
      "max-classes-per-file": "error",
      "max-depth": "error",
      "max-len": "off",
      "max-lines": "error",
      "max-lines-per-function": "off",
      "max-nested-callbacks": "error",
      "max-params": "error",
      "max-statements": "warn",
      "max-statements-per-line": "error",
      "multiline-comment-style": "off",
      "new-cap": "error",
      "new-parens": "error",
      "newline-after-var": "error",
      "newline-before-return": "error",
      "newline-per-chained-call": "error",
      "no-alert": "error",
      "no-array-constructor": "error",
      "no-await-in-loop": "error",
      "no-bitwise": "error",
      "no-buffer-constructor": "error",
      "no-caller": "error",
      "no-catch-shadow": "error",
      "no-confusing-arrow": "error",
      "no-console": "error",
      "no-constructor-return": "error",
      "no-continue": "error",
      "no-div-regex": "error",
      "no-dupe-else-if": "error",
      "no-duplicate-imports": "error",
      "no-else-return": "error",
      "no-empty-function": "off",
      "no-eq-null": "error",
      "no-eval": "error",
      "no-extend-native": "error",
      "no-extra-bind": "error",
      "no-extra-label": "error",
      "no-extra-parens": "error",
      "no-floating-decimal": "error",
      "no-implicit-coercion": "error",
      "no-implicit-globals": "error",
      "no-implied-eval": "error",
      "no-import-assign": "error",
      "no-inline-comments": "error",
      "no-invalid-this": "off",
      "no-iterator": "error",
      "no-label-var": "error",
      "no-labels": "error",
      "no-lone-blocks": "error",
      "no-lonely-if": "error",
      "no-loop-func": "error",
      "no-magic-numbers": "warn",
      "no-mixed-operators": "error",
      "no-mixed-requires": "error",
      "no-multi-assign": "error",
      "no-multi-spaces": "error",
      "no-multi-str": "error",
      "no-multiple-empty-lines": "error",
      "no-native-reassign": "error",
      "no-negated-condition": "error",
      "no-negated-in-lhs": "error",
      "no-nested-ternary": "error",
      "no-new": "off",
      "no-new-func": "error",
      "no-new-object": "error",
      "no-new-require": "error",
      "no-new-wrappers": "error",
      "no-octal-escape": "error",
      "no-param-reassign": "error",
      "no-path-concat": "error",
      "no-plusplus": "error",
      "no-process-env": "error",
      "no-process-exit": "error",
      "no-proto": "error",
      "no-restricted-globals": "error",
      "no-restricted-imports": "error",
      "no-restricted-modules": "error",
      "no-restricted-properties": "error",
      "no-restricted-syntax": "error",
      "no-return-assign": "error",
      "no-return-await": "error",
      "no-script-url": "error",
      "no-self-compare": "error",
      "no-sequences": "error",
      "no-setter-return": "error",
      "no-shadow": "error",
      "no-spaced-func": "error",
      "no-sync": "error",
      "no-tabs": "error",
      "no-template-curly-in-string": "error",
      "no-ternary": "off",
      "no-throw-literal": "error",
      "no-trailing-spaces": "error",
      "no-undef-init": "error",
      "no-undefined": "off",
      "no-underscore-dangle": "error",
      "no-unmodified-loop-condition": "error",
      "no-unneeded-ternary": "error",
      "no-unused-expressions": "error",
      "no-use-before-define": "error",
      "no-useless-call": "error",
      "no-useless-computed-key": "error",
      "no-useless-concat": "error",
      "no-useless-constructor": "error",
      "no-useless-rename": "error",
      "no-useless-return": "error",
      "no-var": "error",
      "no-void": "error",
      "no-warning-comments": "error",
      "no-whitespace-before-property": "error",
      "nonblock-statement-body-position": "error",
      "object-curly-newline": "error",
      "object-curly-spacing": "off",
      "object-property-newline": "error",
      "object-shorthand": "error",
      "one-var": "error",
      "one-var-declaration-per-line": "error",
      "operator-assignment": "error",
      "operator-linebreak": "error",
      "padded-blocks": "off",
      "padding-line-between-statements": "error",
      "prefer-arrow-callback": "error",
      "prefer-const": "error",
      "prefer-destructuring": "off",
      "prefer-exponentiation-operator": "error",
      "prefer-named-capture-group": "error",
      "prefer-numeric-literals": "error",
      "prefer-object-spread": "error",
      "prefer-promise-reject-errors": "error",
      "prefer-reflect": "error",
      "prefer-regex-literals": "error",
      "prefer-rest-params": "error",
      "prefer-spread": "error",
      "prefer-template": "error",
      "quote-props": "off",
      "quotes": "off",
      "radix": "error",
      "require-await": "error",
      "require-jsdoc": "error",
      "require-unicode-regexp": "error",
      "rest-spread-spacing": [
          "error",
          "never"
      ],
      "semi": "off",
      "semi-spacing": "error",
      "semi-style": [
          "error",
          "last"
      ],
      "sort-keys": "off",
      "sort-vars": "off",
      "space-before-blocks": "error",
      "space-before-function-paren": "off",
      "space-in-parens": [
          "error",
          "never"
      ],
      "space-infix-ops": "error",
      "space-unary-ops": "error",
      "spaced-comment": [
          "error",
          "always"
      ],
      "strict": "error",
      "switch-colon-spacing": "error",
      "symbol-description": "error",
      "template-curly-spacing": [
          "error",
          "never"
      ],
      "template-tag-spacing": [
          "error",
          "never"
      ],
      "unicode-bom": [
          "error",
          "never"
      ],
      "valid-jsdoc": "error",
      "vars-on-top": "error",
      "wrap-iife": "error",
      "wrap-regex": "error",
      "yield-star-spacing": "error",
      "yoda": [
          "error",
          "never"
      ]
  }
};


src/swipe.js
export default class Swipe {
  constructor(toast, destroy) {
    this.destroy = destroy;
    this.xDown = null;
    this.zero = 0;
    toast.addEventListener("touchstart", this.handleTouchStart.bind(this), false);
    toast.addEventListener("touchmove", this.handleTouchMove.bind(this), false);
  }

  getTouches(evt) {
    // browser API
    return evt.touches ||
      // jQuery
      evt.originalEvent.touches;
  }

  handleTouchStart(evt) {
    const firstTouch = this.getTouches(evt)[this.zero];

    this.xDown = firstTouch.clientX;
  }

  handleTouchMove(evt) {
    if (!this.xDown) {
      return;
    }

    const xUp = evt.touches[this.zero].clientX,
     xDiff = this.xDown - xUp;

    if (xDiff < this.zero) {
      this.destroy(xDiff);
    } else {
      this.destroy(xDiff);
    }

    this.xDown = null;
  }
}


scripts/queue.js
import Swipe from "./swipe";
const html = String.raw;

const style = html`
  <style>
    @keyframes entrance {
      0% {
        transform: translateY(calc(20px + 100%));
      }
      35% {
        transform: translateY(calc(20px - 15%));
      }
      60% {
        transform: translateY(calc(20px + 8%));
      }
      100% {
        transform: translateY(0%);
      }
    }

    @keyframes exit {
      from {
        opacity: 1;
        transform: translateY(0%);
      }
      to {
        opacity: 0;
        transform: translateX(-100%);
      }
    }

    .toastContainer {
      max-width: 400px;
      width: 100%;
      position: fixed;
      bottom: 20px;
      left: 0;
    }

    @media screen and (min-width: 420px) {
      .toastContainer {
        left: 20px;
      }
    }

    ods-toast {
      transform: translateY(calc(20px + 100%));
      display: inline-block;
    }

    ods-toast.show {
      animation-name: entrance;
      animation-duration: 2s;
      transform: translateY(0%);
    }

    ods-toast.exit {
      animation-name: exit;
      animation-duration: 1500ms;
    }

    @media screen and (max-width: 420px) {
      ods-toast {
        width: 100%;
      }
    }
  </style>
`;
document.body.appendChild(createElementFromHTML(style));

function createElementFromHTML(htmlString) {
  var div = document.createElement("div");
  div.innerHTML = htmlString.trim();

  // Change this to div.childNodes to support multiple top-level nodes
  return div.firstChild;
}

export default class QueueToasts {
  constructor() {
    this.toasts = [];
    this.container = document.createElement("div");
    document.body.appendChild(this.container);
    this.container.className = "toastContainer";
  }

  add(title, message) {
    const toast = document.createElement("ods-toast");
    toast.setAttribute("title", title);
    toast.setAttribute("message", message || "");
    toast.onDestroy = this.destroyCurrentToast.bind(this);
    toast.onClick = this.setAutoDismissal.bind(this);

    this.toasts.push(toast);
    if (this.toasts.length == 1) {
      this.showNextToast();
    }
  }

  showNextToast() {
    if (this.toasts.length > 0) {
      const currentToast = this.toasts[0];
      this.container.appendChild(currentToast);
      currentToast.classList.add("show");
      new Swipe(currentToast, this.destroyCurrentToast.bind(this));
      this.setAutoDismissal();
    }
  }

  destroyCurrentToast() {
    console.log(this.toasts.length);
    if (this.toasts.length) {
      const currentToast = this.toasts[0];
      currentToast.classList.remove("show");
      currentToast.classList.add("exit");
      setTimeout(() => {
        this.container.removeChild(currentToast);
        this.toasts.shift();
        this.showNextToast();
      }, 1500);
    }
  }

  setAutoDismissal() {
    if (!!this.timeoutHandle) {
      clearTimeout(this.timeoutHandle);
    }
    this.timeoutHandle = setTimeout(this.destroyCurrentToast.bind(this), 8000);
  }
}

const toasts = new QueueToasts();
toasts.add("1", "you done good! way to go!");
toasts.add("2");


src/ods-toast.js
// Copyright (c) Alaska Air. All right reserved. Licensed under the Apache-2.0 license
// See LICENSE in the project root for license information.

// ---------------------------------------------------------------------

import { LitElement, html } from "lit-element";

// Import touch detection lib
import "focus-visible/dist/focus-visible.min.js";
// import the processed CSS file into the scope of the component
import componentProperties from "./tokens/componentShapeProperties-css.js";
import styleCss from "./style-css.js";
import iconProperties from "@alaskaairux/orion-icons/dist/tokens/CSSTokenProperties-css.js";
import checkmark from "@alaskaairux/orion-icons/dist/icons/interface/classic-checkmark_es6";
import close from "@alaskaairux/orion-icons/dist/icons/close_es6";

// build the component class
class OdsToast extends LitElement {
  constructor() {
    super();
    this.onDestroy = null;
    this.onClick = null;

    this.dom = new DOMParser().parseFromString(checkmark.svg, "text/html");
    this.checkmarksvg = this.dom.body.firstChild;

    this.dom = new DOMParser().parseFromString(close.svg, "text/html");
    this.closesvg = this.dom.body.firstChild;
  }

  // function to define props used within the scope of thie component
  static get properties() {
    return {
      title:          { type: String },
      message:        { type: String },
      clickCallback:  { type: Function }
    };
  }

  destroyCallback() {
    this.parentElement.removeChild(this);
  }

  clickCallback() {}

  // function that renders the HTML and CSS into the scope of the component
  render() {
    return html`
      ${componentProperties}
      ${iconProperties}
      ${styleCss}

      <div class="ods-toast" aria-role="status" aria-live="assertive"  @click=${this.clickCallback.bind(this)}>

        <div class="primaryContent" @click=${this.destroyCallback.bind(this)}>
          <div class="primaryContent-container primaryContent-container--icon">
            ${this.checkmarksvg}
          </div>
          <div class="primaryContent-container primaryContent-container--text">
            <div class="primaryContent-title">${this.title}</div>
            <div class="primaryContent-message">${this.message}</div>
            <div class="secondaryContent">
              <slot></slot>
            </div>
          </div>
          <div class="primaryContent-container primaryContent-container--exit" alt="Close">
            ${this.closesvg}
          </div>
        </div>
      </div>
    `;
  }
}

customElements.define("ods-toast", OdsToast);


src/toaster.js
export default class Toaster {
  constructor(displayTime) {
    this.toasts = [];
    this.waitTime = 7000;
    this.nextToast = 300;
    this.zero = 0;
    this.one = 1;
    this.esc = 27;
    this.return = 13;
    this.displayTime = displayTime || this.waitTime;
    this.container = document.createElement("div");
    this.container.setAttribute("aria-live", "polite");
    document.body.appendChild(this.container);
    this.container.className = "ods-toast__toastContainer";
  }

  add(title, message, slotHTML) {
    const toast = document.createElement("ods-toast");

    toast.setAttribute("title", title);
    toast.setAttribute("message", message || "");
    toast.innerHTML = slotHTML || "";
    toast.destroyCallback = this.destroyCurrentToast.bind(this);
    toast.clickCallback = this.setAutoDismissal.bind(this);

    this.toasts.push(toast);
    if (this.toasts.length === this.one) {
      this.showNextToast();
    }
  }

  showNextToast() {
    if (this.toasts.length > this.zero) {
      const currentToast = this.toasts[this.zero];

      this.container.appendChild(currentToast);
      currentToast.classList.add("ods-toast__showToast");
      currentToast.tabIndex = this.zero;
      currentToast.addEventListener("keydown", (ex) => {

        // Supports either the esc or return keys
        if (ex.keyCode === this.esc || ex.keyCode === this.return) {
          this.destroyCurrentToast();
        }
      });

      /* global Swipe */

      if (typeof Swipe !== "undefined") {
        new Swipe(currentToast, this.destroyCurrentToast.bind(this));
      }

      this.setAutoDismissal();
    }
  }

   // animation-duration(css) must match _destroyCurrentToast() setTimeout value
  destroyCurrentToast(xDiff) {
    if (this.toasts.length) {
      const currentToast = this.toasts[this.zero];

      currentToast.classList.remove("ods-toast__showToast");

      if (xDiff < this.zero) {
        currentToast.classList.add("ods-toast__exitToast--right");
      } else if (xDiff > this.zero) {
        currentToast.classList.add("ods-toast__exitToast--left");
      } else {
        currentToast.classList.add("ods-toast__exitToast--down");
      }

      setTimeout(() => {
        this.container.removeChild(currentToast);
        this.toasts.shift();
        this.showNextToast();
      }, this.nextToast);
    }
  }

  setAutoDismissal() {
    if (this.timeoutHandle) {
      clearTimeout(this.timeoutHandle);
    }
    this.timeoutHandle = setTimeout(
      this.destroyCurrentToast.bind(this),
      this.displayTime
    );
  }
}
